AssertionError: AbstractLinkable._notifier is None in _notify_links method

* gevent version: "~21.12.0" pypi
* Python version: python:3.9 from slim-buster docker image
* Operating System: python:3.9-slim-buster docker image

### Description:

Running django 3.2 using gunicorn 20.1. I keep seeing this traceback in my logs from `gevent/_abstract_linkable.py` `assert self._notifier is notifier, (self._notifier, notifier)`. 

```python-traceback
Traceback (most recent call last):
File "src/gevent/_abstract_linkable.py", line 287, in gevent._gevent_c_abstract_linkable.AbstractLinkable._notify_links
File "src/gevent/_abstract_linkable.py", line 333, in gevent._gevent_c_abstract_linkable.AbstractLinkable._notify_links
AssertionError: (None, <callback at 0x7f3af9646600 args=([],)>)
```

No other logs surrounding these in the given context.

Is this something to be concerned about or anything I can fix?

### What I've run:
`ddtrace-run gunicorn --workers 5 --worker-class gevent project_name.wsgi:application`

Digging into this a bit more, I've found that when I start my application with

```
ddtrace-run gunicorn -k gevent main:app
```

gevent version: "~21.12.0" pypi
Python version: python:3.9 from slim-buster docker image
Operating System: python:3.9-slim-buster docker image

I see these errors emitted at the beginning

```
Traceback (most recent call last):
File "src/gevent/_abstract_linkable.py", line 287, in gevent._gevent_c_abstract_linkable.AbstractLinkable._notify_links
File "src/gevent/_abstract_linkable.py", line 333, in gevent._gevent_c_abstract_linkable.AbstractLinkable._notify_links
AssertionError: (None, <callback at 0x7f3af9646600 args=([],)>)
```

but the server actually does startup and handles http requests in an asynchronous way. I tried load testing this by sending ~1000 requests to an endpoint that `time.sleep()`s for 5 seconds and another that makes an http request to a server that sleeps for five seconds and in both cases all 1000 requests returned in 5 seconds. So it seems like these errors are not actually affecting the performance of the server. This is by no means comprehensive and it may be that certain parts of the library are broken but the ones I tested work. But based on my limited experience it seems that things are still working. It would be great to understand what is causing this error and understand its broader scope

--- a/docs/changes/1895.bugfix.rst
+++ b/docs/changes/1895.bugfix.rst
@@ -1,0 +1,5 @@
+Prevent an ``AssertionError`` (from ``AbstractLinkable``, such as
+locks, events, etc) from being printed after ``os.fork`` under certain
+conditions.
+
+See also :issue:`2058`.

--- a/src/gevent/_abstract_linkable.py
+++ b/src/gevent/_abstract_linkable.py
@@ -1,13 +1,16 @@
 # -*- coding: utf-8 -*-
 # cython: auto_pickle=False,embedsignature=True,always_allow_keywords=False
 """
-Internal module, support for the linkable protocol for "event" like objects.
+Internal use only base class for objects
+that can be waited on and then asked to notify (wake up)
+waiters.  Examples include locks, events, semaphores.
 
+gevent has the generic concept of "linkable" objects, or the "linkable"
+protocol, which is the ``link`` and ``rawlink`` methods. (Actually sending
+the notification is up to the implementation of the individual
+objects.) `gevent.greenlet.Greenlet` implements this protocol but
+does not extend this object (TODO: It probably should.)
 """
-from __future__ import absolute_import
-from __future__ import division
-from __future__ import print_function
-
 import sys
 from gc import get_objects
 
@@ -330,7 +333,12 @@ class AbstractLinkable(object):
             # We should not have created a new notifier even if callbacks
             # released us because we loop through *all* of our links on the
             # same callback while self._notifier is still true.
-            assert self._notifier is notifier, (self._notifier, notifier)
+            # However, one of our callbacks could have called ``os.fork``,
+            # which invokes ``_at_fork_reinit`` and destroys the notifier and
+            # hub *while this method is running*.
+            assert self._notifier is notifier or (
+                self._notifier is None and self.hub is None
+            ), (self, self._notifier, notifier)
             self._notifier = None
             # TODO: Maybe we should intelligently reset self.hub to
             # free up thread affinity? In case of a pathological situation where

--- a/src/gevent/event.py
+++ b/src/gevent/event.py
@@ -83,9 +83,10 @@ class Event(AbstractLinkable): # pylint:disable=undefined-variable
         self._flag = False
 
     def __str__(self):
-        return '<%s.%s %s _links[%s]>' % (
+        return '<%s.%s at 0x%x %s _links[%s]>' % (
             self.__class__.__module__,
             self.__class__.__name__,
+            id(self),
             'set' if self._flag else 'clear',
             self.linkcount()
         )

--- a/src/gevent/exceptions.py
+++ b/src/gevent/exceptions.py
@@ -6,9 +6,6 @@ Exceptions.
 .. versionadded:: 1.3b1
 
 """
-from __future__ import absolute_import
-from __future__ import division
-from __future__ import print_function
 
 from greenlet import GreenletExit
 

--- a/src/gevent/monkey/_patch_thread_common.py
+++ b/src/gevent/monkey/_patch_thread_common.py
@@ -7,11 +7,13 @@ Internal use only.
 """
 import sys
 
-from ._util import _patch_module
-from ._util import _queue_warning
-from ._util import _notify_patch
+from gevent.exceptions import LoopExit
 
 from ._state import is_object_patched
+from ._util import _notify_patch
+from ._util import _patch_module
+from ._util import _queue_warning
+
 
 def _patch_existing_locks(threading):
     if len(list(threading.enumerate())) != 1:
@@ -191,8 +193,9 @@ class BasePatcher:
             self.patch_logging()
 
     def patch_event(self):
-        from .api import patch_item
         from gevent.event import Event
+
+        from .api import patch_item
         patch_item(self.threading_mod, 'Event', Event)
         # Python 2 had `Event` as a function returning
         # the private class `_Event`. Some code may be relying
@@ -215,8 +218,9 @@ class BasePatcher:
 
     def patch__threading_local(self):
         _threading_local = __import__('_threading_local')
-        from .api import patch_item
         from gevent.local import local
+
+        from .api import patch_item
         patch_item(_threading_local, 'local', local)
 
     def patch_active_threads(self):
@@ -266,14 +270,19 @@ class BasePatcher:
             # XXX: There's probably a better way to do this. Probably need to take a
             # step back and look at the whole picture.
             main_thread._ident = get_ident()
-            orig_shutdown()
+            try:
+                orig_shutdown()
+            except LoopExit: # pragma: no cover
+                pass
             patch_item(threading_mod, '_shutdown', orig_shutdown)
         patch_item(threading_mod, '_shutdown', _shutdown)
 
     @staticmethod # Static to be sure we don't accidentally capture `self` and keep it alive
     def _make_existing_non_main_thread_join_func(thread, thread_greenlet, threading_mod):
-        from gevent.hub import sleep
         from time import time
+
+        from gevent.hub import sleep
+
         # TODO: This is almost the algorithm that the 3.13 _ThreadHandle class
         # employs. UNIFY them.
         def join(timeout=None):

--- a/src/gevent/monkey/_patch_thread_gte313.py
+++ b/src/gevent/monkey/_patch_thread_gte313.py
@@ -7,8 +7,10 @@ Internal use only.
 """
 import sys
 
+from gevent.exceptions import LoopExit
 
 from ._patch_thread_common import BasePatcher
+
 
 class Patcher(BasePatcher):
 
@@ -18,8 +20,8 @@ class Patcher(BasePatcher):
 
         for thread in self.threading_mod._active.values():
             if thread == main_native_thread():
-                from greenlet import getcurrent
                 from gevent.thread import _ThreadHandle
+                from greenlet import getcurrent
                 thread._after_fork = lambda new_ident=None: new_ident
                 handle = _ThreadHandle()
                 handle._set_greenlet(getcurrent())
@@ -36,6 +38,7 @@ class Patcher(BasePatcher):
 
     def patch_threading_shutdown_on_main_thread_not_already_patched(self):
         import greenlet
+
         from .api import patch_item
 
         main_thread = self.main_thread
@@ -56,7 +59,7 @@ class Patcher(BasePatcher):
                 # A greenlet could have .kill() us
                 # or .throw() to us. I'm the main greenlet,
                 # there's no where else for this to go.
-                from gevent  import get_hub
+                from gevent import get_hub
                 get_hub().print_exception(_greenlet, *sys.exc_info())
 
             # Now, this may have resulted in us getting stopped
@@ -93,7 +96,10 @@ class Patcher(BasePatcher):
                         handle._set_done()
                         break
 
-            orig_shutdown()
+            try:
+                orig_shutdown()
+            except LoopExit: # pragma: no cover
+                pass
             patch_item(threading_mod, '_shutdown', self.orig_shutdown)
 
         patch_item(self.threading_mod, '_shutdown', _shutdown)

--- a/src/gevent/monkey/_patch_thread_lt313.py
+++ b/src/gevent/monkey/_patch_thread_lt313.py
@@ -6,7 +6,7 @@ prior to 3.13.
 Internal use only.
 """
 import sys
-
+from gevent.exceptions import LoopExit
 
 from ._patch_thread_common import BasePatcher
 
@@ -72,7 +72,10 @@ class Patcher(BasePatcher):
             # acquire should be our own (hopefully), and the call to
             # _stop that orig_shutdown makes will discard it.
 
-            orig_shutdown()
+            try:
+                orig_shutdown()
+            except LoopExit: # pragma: no cover
+                pass
             patch_item(threading_mod, '_shutdown', orig_shutdown)
 
         patch_item(threading_mod, '_shutdown', _shutdown)

--- a/src/gevent/tests/test__lock.py
+++ b/src/gevent/tests/test__lock.py
@@ -1,7 +1,3 @@
-from __future__ import absolute_import
-from __future__ import division
-from __future__ import print_function
-
 from gevent import lock
 
 
@@ -28,6 +24,55 @@ class TestRLockMultiThread(test__semaphore.TestSemaphoreMultiThread):
         self.assertIsNone(sem._block.hub)
 
 
+class TestLockReinitAfterFork(greentest.TestCase):
+
+    def test_it(self):
+        # See https://github.com/gevent/gevent/issues/1895
+
+        # Make sure we carefully handle forking while running callbacks
+        # of a lock/_AbstractLinkable object.
+        import sys
+        import subprocess
+        import textwrap
+        import os
+
+        if not hasattr(os, 'fork'):
+            self.skipTest('Requires os.fork')
+
+        script =textwrap.dedent("""\
+        from gevent import monkey
+        monkey.patch_all()
+
+        from gevent import spawn
+
+        import os
+        from logging import Handler
+        from threading import Event
+        handler = Handler()
+        handler.acquire()
+
+        event = Event()
+
+        def forker():
+            event.set()
+            handler.acquire()
+            handler.release()
+            os.fork()
+
+        g = spawn(forker)
+        event.wait()
+        handler.release()
+        g.join()
+        """)
+
+        output = subprocess.check_output(
+            [sys.executable, '-c', script],
+            stderr=subprocess.STDOUT,
+        ).decode('utf-8')
+
+        # We used to print a failing AssertionError to stdout,
+        # that should no longer be the case.
+        self.assertFalse(output)
 
 if __name__ == '__main__':
     greentest.main()

